streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorTopology.java

```diff
@@ -59,4 +78,78 @@ public SourceNode source(String topic) {
         return stateStoreSuppliers;
     }
 
+
+    /**
+     * Produces a string representation contain useful information about a ProcessorTopology.
+     * This is useful in debugging scenarios. Prints graph by performing a breadth-first search starting
+     * from source nodes
+     * @return A string representation of the ProcessorTopology instance.
+     */
+    public String toString() {
+        Map<String, Boolean> visitedNode = new HashMap<String, Boolean>();
+        Queue queue = new LinkedList();
+        String processorTopologyString = "ProcessorTopology[";
+        Set<SourceNode> tmpSources = sources();
+
+        // pre-process source nodes to get reverse mapping
+        Map<String, String> sourceNameToTopic = new HashMap<>();
+        for (String topic : sourceByTopics.keySet()) {
+            SourceNode source = sourceByTopics.get(topic);
+            sourceNameToTopic.put(source.name(), topic);
+        }
+
+        // pre-process sink nodes to get reverse mapping
+        Map<String, String> sinkNameToTopic = new HashMap<>();
+        for (String topic : sinkByTopics.keySet()) {
+            SinkNode sink = sinkByTopics.get(topic);
+            sinkNameToTopic.put(sink.name(), topic);
+        }
+
+        // add all sources to a queue for Breadth-First traversal
+        if (tmpSources != null) {
+            // add all sources
+            for (String topic : sourceByTopics.keySet()) {
+                SourceNode source = sourceByTopics.get(topic);
+                // mark as visited
+                visitedNode.put(source.name(), true);
+                // put in queue
+                queue.add(source);
+            }
+
+            // iterate to next level
```

Comment:
I feel there are still some room for improvements. @theduderog's feedbacks would be very helpful, while here are my two cents:

1. Start a new line with different levels of indent as we go through the streams granularities, streams->threads->tasks->topology, etc.

2. For the topology specifically, the parent-child relationship seems incorrect from the `-->` hierarchy, so I'd suggest reducing the verboseness a bit by removing `sourceNode[..., sourceTopic,` etc, and also just displays the child name instead of using `-->`, for example take the above topology:

```
KSTREAM-SOURCE-0000000000: topics [streams-file-input], children [KSTREAM-FLATMAPVALUES-0000000001]
KSTREAM-FLATMAPVALUES-0000000001: children [KSTREAM-MAP-0000000002]
KSTREAM-MAP-0000000002: children [KSTREAM-AGGREGATE-0000000003]
KSTREAM-AGGREGATE-0000000003: stateStores [Counts], children [name=KSTREAM-FILTER-0000000005]
KSTREAM-FILTER-0000000005: children [KSTREAM-SINK-0000000004]
KSTREAM-SINK-0000000004: topic [X-KSTREAM-MAP-0000000002-repartition], children [KSTREAM-FILTER-0000000005]

KSTREAM-SOURCE-0000000006: topics [X-KSTREAM-MAP-0000000002-repartition], children [KTABLE-TOSTREAM-0000000007]
KTABLE-TOSTREAM-0000000007: children [KSTREAM-SINK-0000000008]
KSTREAM-SINK-0000000008: topics [streams-wordcount-output]
```